### ISSUE TITLE
[Tracking] ONNX 1.9 work items

### ISSUE THREAD

**Issue by askhade** [onnx/onnx#3225]
Creating this issue to collect and track all the work items for ONNX 1.9 release.
The release is planned for March 2021.

@onnx/sig-archinfra , @onnx/sig-operators , @onnx/steering-committee , @sveta-levitan, @wschin , @chinhuang007 and community members, please add work items here so that we can scope the release.

Thanks!

Currently planned workitems for 1.9

1. Removing Optimizers from onnx packages @daquexian 
2. Updates to checker and shape inference exception handling @askhade
3. Selective schema loading @jcwchen 
4. Opset 14 updates
5. Updates to external data helpers (add more options to control which tensors are serialized, add a new API to convert and serialize model in 1 step as opposed to the 2 steps required now) @annajung 
6. Enhance ONNX type-checker: currently it does not check type-constraints specified in the op-schema spec @jcwchen
7. Enhance the sparse-tensor constants to allow block-sparse format? <if time permits; needs approval from community>
8. Should we migrate sparse-tensor type from ONNXML to ONNX? <Needs agreement from community.>

**Comment by spandantiwari** [comment]
For opset 14, two possibilities to consider are:
1) Adding `trilu` op to spec.
2) Standardize specification of optional inputs/outputs.

**Comment by askhade** [comment]
@spandantiwari : Thanks for the inputs. Can you elaborate more on your 2nd bullet... I don't understand what is the exact work that needs to go in for this item.

**Comment by gramalingam** [comment]
Some possible enhancements (though not sure whether it should be in 1.9; depends on workload and prioritization).
1) Enhance ONNX type-checker: currently it does not check type-constraints specified in the op-schema spec (unlike ORT).
2) Enhance the sparse-tensor constants to allow block-sparse format?
3) Should we migrate sparse-tensor type from ONNXML to ONNX?

**Comment by jcwchen** [comment]
There is still a IR gap issue for version_converter:
https://github.com/onnx/onnx/issues/2873. Anyone interested is welcome.

**Comment by daquexian** [comment]
Is there any plan to support python 3.9 in onnx 1.9 release?

**Comment by neginraoof** [comment]
Adding Trilu operator: https://github.com/onnx/onnx/pull/3291

**Comment by BowenBao** [comment]
Extending type support for math ops(Add, Sub, Mul, Div) #3334 (for HuggingFace/transfo_xl)

**Comment by jcwchen** [comment]
- Add weekly CI: provide ONNX TestPyPI packages for verification https://github.com/onnx/onnx/pull/3283
- Selective load ONNX schema by specific opset_version https://github.com/onnx/onnx/pull/3266

@daquexian According to the discussion with release manger of ONNX 1.9, Python 3.9 support is in the plan. Thanks.

**Comment by etusien** [comment]
# Key Updates
 
- Removed Optimizers from ONNX packages https://github.com/onnx/onnx/pull/3288
- Selective schema loading by specific `opset_version` https://github.com/onnx/onnx/pull/3266
- Updates to external data helpers (more options to control which attributes are converted to external data and whether a model should be saved) https://github.com/onnx/onnx/pull/3280
- New adapter for opset version converter https://github.com/onnx/onnx/pull/3343

## Opset version 14 

- New operator:
	- [HardSwish](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#HardSwish-14) https://github.com/onnx/onnx/pull/3332
	- [Trilu](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Trilu-14) https://github.com/onnx/onnx/pull/3291
- Extended supported types to include uint8, int8, uint16, and int16. https://github.com/onnx/onnx/pull/3334
	- [Add](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Add-14)
	- [Div](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Div-14)
	- [Mul](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Mul-14)
	- [Sub](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Sub-14)
- Added `allowzero` attribute to [Reshape](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Reshape-14) operator https://github.com/onnx/onnx/pull/3113
- Allowed recurrent operations to be batchwise  https://github.com/onnx/onnx/pull/3217
	- [GRU](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#GRU-14)
	- [LSTM](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#LSTM-14)
	- [RNN](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#RNN-14)
- Expanded [Cum-Sum](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#CumSum-14) to float16 and bfloat16 data types https://github.com/onnx/onnx/pull/3195
- Expanded [Relu](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Relu-14) to all signed data types https://github.com/onnx/onnx/pull/3141
- Added training-mode support to [BatchNorm](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#BatchNorm-14)  https://github.com/onnx/onnx/pull/3333

## API
- `onnx.OnnxParser`  Parser for a textual syntax of ONNX models https://github.com/onnx/onnx/pull/3194 

## Infrastructure
- Removed Python 3.5 in all release pipelines https://github.com/onnx/onnx/pull/3353
- Added Python 3.9 to all release pipelines https://github.com/onnx/onnx/pull/3352
- Added weekly CI: provide ONNX TestPyPI packages for verification #3283
- Reduced binary size for Linux and Mac package https://github.com/onnx/onnx/pull/3337
- Added check for uploaded/generated backend test data in release CIs https://github.com/onnx/onnx/pull/3274 
- Enabled no exception build and updated exception handling for DataTypeUtils
https://github.com/onnx/onnx/pull/3265

## Bug fixes 
- Added missing test for ConvInteger without padding https://github.com/onnx/onnx/pull/3288 
- Updated Resize op test to opset 13 
- Expanded `Ir_pb_converter` to empty shape https://github.com/onnx/onnx/pull/3279

# Notes 
- Be aware of protobuf version gap issue (like building onnx with protobuf>=3.12 is not compatible with older protobuf)

**Comment by etusien** [comment]
In the comment above, I proposed notes regarding the upcoming 1.9 release. Please let me know about any mistake or missing scope item. Thanks!
